integration_test_suite/integration_tests/src/main/java/io/prometheus/jmx/test/httpserver/general/BlackListAndWhiteListObjectNames_IT.java

```diff
@@ -0,0 +1,210 @@
+
+        // Get the id to force the network creation
+        network.getId();
+    }
+
+    @TestEngine.BeforeAll
+    public void beforeAll() throws Exception {
+        // Application container
+        applicationContainer = new GenericContainer<>(dockerImageName)
+                .waitingFor(Wait.forLogMessage(".*Running.*", 1))
+                .withClasspathResourceMapping("common", "/temp", BindMode.READ_ONLY)
+                .withClasspathResourceMapping(getClass().getName().replace(".", "/"), "/temp", BindMode.READ_ONLY)
+                .withCommand("/bin/sh application.sh")
+                .withExposedPorts(9999)
+                .withLogConsumer(outputFrame -> System.out.print(outputFrame.getUtf8String()))
+                .withNetwork(network)
+                .withNetworkAliases("application")
+                .withStartupTimeout(Duration.ofMillis(30000))
+                .withStartupCheckStrategy(new IsRunningStartupCheckStrategy())
+                .withWorkingDirectory("/temp");
+
+        if (DockerImageNameParameters.isJava6(dockerImageName)) {
+            applicationContainer.withCommand("/bin/sh application_java6.sh");
+        }
+
+        applicationContainer.start();
+
+        // Exporter container
+        exporterContainer = new GenericContainer<>(dockerImageName)
+                .waitingFor(Wait.forLogMessage(".*Running.*", 1))
+                .withClasspathResourceMapping("common", "/temp", BindMode.READ_ONLY)
+                .withClasspathResourceMapping(getClass().getName().replace(".", "/"), "/temp", BindMode.READ_ONLY)
+                .withCommand("/bin/sh exporter.sh")
+                .withExposedPorts(8888)
+                .withLogConsumer(outputFrame -> System.out.print(outputFrame.getUtf8String()))
+                .withNetwork(network)
+                .withNetworkAliases("exporter")
+                .withStartupCheckStrategy(new IsRunningStartupCheckStrategy())
+                .withWorkingDirectory("/temp");
+
+        if (DockerImageNameParameters.isJava6(dockerImageName)) {
+            exporterContainer.withCommand("/bin/sh exporter_java6.sh");
+        }
+
+        exporterContainer.start();
+
+        // HTTP client
+        httpClient = new HttpClient("http://localhost:" + exporterContainer.getMappedPort(8888));
+    }
+
+    @TestEngine.Test
+    public void testHealthy() throws Exception {
+        String path = "/-/healthy";
+        Request.Builder requestBuilder = httpClient.createRequest(path);
+        try (Response response = httpClient.execute(requestBuilder)) {
+            assertThat(response).isNotNull();
+            assertThat(response.code()).isEqualTo(200);
+            ResponseBody responseBody = response.body();
+            assertThat(responseBody).isNotNull();
+            String content = responseBody.string();
+            assertThat(content).isNotNull();
+            assertThat(content).isEqualTo("Exporter is Healthy.");
+        }
+    }
```

Comment:
This test (and others below) are copied to all test files. It would be good if the test was implemented only once, and then run with different parameters.

Reply:
Definitely can refactor... though there are issues with inheritance using IntelliJ.

Example:

```
public abstract class BaseTest {

    @Test
    protected void testA() {
        // test code
    }

}
```

```
public class ConcreteTest extends BaseTest {
    // no methods, we want to use the methods in BaseTest
}
```

There are no members to run from IntelliJ in the **Project** view. The class (classname) is really a _marker_ for the configuration to use the base class.

Switch to the **Structure** view in IntelliJ for the class and enable the view to show inherited methods

![image](https://user-images.githubusercontent.com/19376470/229975229-e7172750-32ea-4b06-bd9d-5b6d676ec829.png)

The method is now listed and has a right-click **Run** option... but it doesn't run because it's a method in the abstract class `BaseTest`

I debugged the JUnit `engineDiscoveryRequest` which confirms IntelliJ is trying to run the method against the abstract base class (This is regardless of the test engine being used)

![image](https://user-images.githubusercontent.com/19376470/229974638-7146c84a-b5a7-4d6d-b5e3-77bb6a546124.png)

If the `BaseTest` is not abstract, it will run... but will run in the context of the `BaseTest` and not ` ConcreteTest` defeating the intent of trying to run the tests in `ConcreteTest`

---

Another approach...

Example:

```
public abstract class BaseTest {

    protected void testA() {
        // test code
    }

}
```

```
public class ConcreteTest extends BaseTest {

    @Test
    public void testA() {
        super.testA();
}
```

This will work and is less code in the test classes, but results in `BaseTest` being a "God Class" for all tests (or a set of God classes) which is an anti-pattern.

It would allow something like...

public abstract class BaseTest {

    protected void testHealthy(String username, String password, int expectedResponseCode) {
        // test code
    }

}

```
public class MyTest extends BaseTest {

    @Test
    public void testHealthy() {
        super.testHealty(String username, String password, int expectedResponseCode);
}
```

... or using functional interfaces...

```
public class UsernamePassword {

    public static Optional<UsernamePassword> of(String username, String password) {
        return Optional.of(new UsernamePassword(username, password));
    }
```

```
public abstract class BaseTest {

    protected void testHealthy(Optional<UsernamePassword> optional, Consumer<Integer> consumer) {
        // test code
    }

}
```

```
public class MyTest extends BaseTest {

    @Test
    public void testHealthy() {
        super.testHealty(
             UsernamePassword.of("prometheus", "test"),
             responseCode -> assertThat(responseCode).isEqualTo(200));
}
```

(Use of functional classes/interfaces is my preference between two above)

Reply:
There's no need to use inheritance. What do you think of a helper class with static methods like this:
```java
public class TestUtil {
    public static void runHealthTest(HttpClient httpClient, String username, String password) throws Exception {
        String path = "/-/healthy";
        Request.Builder requestBuilder = httpClient.createRequest(path);
        requestBuilder.addHeader(HttpHeader.AUTHORIZATION, basicAuthentication(username, password));
        try (Response response = httpClient.execute(requestBuilder)) {
            assertThat(response).isNotNull();
            assertThat(response.code()).isEqualTo(200);
            ResponseBody responseBody = response.body();
            assertThat(responseBody).isNotNull();
            String content = responseBody.string();
            assertThat(content).isNotNull();
            assertThat(content).isEqualTo("Exporter is Healthy.");
        }
    }
    // same for other code that's reused in multiple tests
}
```

Now you can call this method in the tests.
```java
    @TestEngine.Test
    public void testHealthy() throws Exception {
        TestUtil.runHealthTest(httpClient, USERNAME, PASSWORD);
    }
```

Reply:
This may be the best solution.

Given...

```
public interface Credentials {

    public void apply(Request.Builder requestBuilder) {
    }
}
```

```
public class BasicAuthenticationCredentials implements Credentials {
    public BasicAuthenticationCredentials(String username, String password) {
    }

    public void apply(Request.Builder requestBuilder) {
    }
}
```

We could test using valid credentials...

```
TestUtil.testHealthy(httpClient, new BasicAuthenticationCredentials(USERNAME, PASSWORD)).expect(200);
```

We could test without credentials...

```
TestUtil.testHealthy(httpClient, null).expect(401);
```

We could test using bad credentials...

```
TestUtil.testHealthy(httpClient, new BasicAuthenticationCredentials(USERNAME, INVALID_PASSWORD).expect(401);
```

---

This pattern would even work for testing the metrics response, which I feel should be test-case specific...

```
TestUtil.testMetrics(httpClient, null).expect(200, Consumer<Response> consumer);
```

```
TestUtil.testPrometheusMetrics(httpClient, null).expect(200, Consumer<Response> consumer);
```

```
TestUtil.testOpenMetrics(httpClient, null).expect(200, Consumer<Response> consumer);
```

... and make the test class implement `Consumer<Response>`

---

Thoughts?

Reply:
Using `TestUtil` as more of a "builder" it would be something like...

```
    @TestEngine.Test
    public void testHealthyWithUsernameWithPassword() throws Exception {
        TestUtil
                .testHealthy(httpClient)
                .withCredentials(new BasicAuthenticationCredentials(USERNAME, PASSWORD))
                .expect(200);
    }

    @TestEngine.Test
    public void testHealthyWithUsernameWithoutPassword() throws Exception {
        TestUtil
                .testHealthy(httpClient)
                .withCredentials(new BasicAuthenticationCredentials(USERNAME, null))
                .expect(401);
    }
```

This feels cleaner to me.

Reply:
@fstab I refactored `io.prometheus.jmx.test.javaagent.general.Minimal_IT` and `io.prometheus.jmx.test.javaagent.general.WhiteListObjectNames_IT` using ideas discussed above.

Take a look.